include/wil/registry_basic.h

```diff
@@ -77,1599 +77,1826 @@ namespace details
+                [[nodiscard]] ::wil::reg::details::value_iterator begin() const
+                {
+                    return ::wil::reg::details::value_iterator(m_hkey, ::wil::reg::details::iterator_creation_flag::begin);
+                }
 
-inline DWORD get_value_dword(HKEY key, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    return get_value_dword(key, nullptr, value_name);
-}
+                [[nodiscard]] ::wil::reg::details::value_iterator end() const WI_NOEXCEPT
+                {
+                    return ::wil::reg::details::value_iterator(m_hkey, ::wil::reg::details::iterator_creation_flag::end);
+                }
 
-inline optional_value<DWORD> try_get_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    const reg_view_details::reg_view regview{key};
-    optional_value<DWORD> value;
-    regview.try_get_value<DWORD>(subkey, value_name, value);
-    return value;
-}
+            private:
+                friend class reg_view_t;
+                // non-owning resource (just get a copy from the owner)
+                HKEY m_hkey;
+            };
 
-inline optional_value<DWORD> try_get_value_dword(HKEY key, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    return try_get_value_dword(key, nullptr, value_name);
-}
+            [[nodiscard]] key_enumerator reg_enum_keys() const
+            {
+                return key_enumerator(::wil::reg::details::get_key(m_key));
+            }
 
-inline DWORD64 get_value_qword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    const reg_view_details::reg_view regview{ key };
-    return regview.get_value<DWORD64>(subkey, value_name);
-}
+            [[nodiscard]] value_enumerator reg_enum_values() const
+            {
+                return value_enumerator(::wil::reg::details::get_key(m_key));
+            }
+#endif // #if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
 
-inline DWORD64 get_value_qword(HKEY key, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    return get_value_qword(key, nullptr, value_name);
-}
+        private:
+            T m_key{};
+        };
 
-inline optional_value<DWORD64> try_get_value_qword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    const reg_view_details::reg_view regview{key};
-    optional_value<DWORD64> value;
-    regview.try_get_value<DWORD64>(subkey, value_name, value);
-    return value;
-}
+        template <typename T, typename err_policy = err_exception_policy>
+        static reg_view_t<T, err_policy> create_reg_view(T&& t)
+        {
+            return reg_view_t<T, err_policy>(::wistd::forward<T>(t));
+        }
 
-inline optional_value<DWORD64> try_get_value_qword(HKEY key, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
-{
-    return try_get_value_qword(key, nullptr, value_name);
-}
+        // reg_view with the raw HKEY type is non-owning
+        using reg_view_nothrow = reg_view_details::reg_view_t<HKEY, ::wil::err_returncode_policy>;

```

Comment:
I propose:

* `unique_reg` & `unique_reg_nothrow` (or `unique_reg_view`, etc)
* `shared_reg` & `shared_reg_nothrow`
* `reg_view` & `reg_view_nothrow`

This aligns with the style for unique & shared ptrs, plus it also separtes reg _views_ which are non-owning and reg "things" which are owning.

It also may still be worth re-examining `unique_hkey` et al and adding all these functions there.